exporter/ststopologyexporter/internal/topology.go

```diff
@@ -43,40 +45,95 @@ func (c *ComponentsCollection) AddResource(attrs *pcommon.Map) bool {
 			ComponentType{
 				"namespace",
 			},
-			newData().
+			newComponentData().
 				withLayer("urn:stackpack:common:layer:applications").
 				withEnvironment(attrs).
 				withName(attrs, "service.namespace"),
 		}
 	}
+
+	serviceIdentifier := fmt.Sprintf("urn:opentelemetry:namespace/%s:service/%s", serviceNamespace.AsString(), serviceName.AsString())
 	c.services = append(c.services, &Component{
-		fmt.Sprintf("urn:opentelemetry:namespace/%s:service/%s", serviceNamespace.AsString(), serviceName.AsString()),
+		serviceIdentifier,
 		ComponentType{
 			"service",
 		},
-		newData().
+		newComponentData().
 			withLayer("urn:stackpack:common:layer:services").
 			withEnvironment(attrs).
 			withName(attrs, "service.name").
 			withVersion(attrs, "service.version").
-			withTag(attrs, "service.namespace"),
+			withTag(attrs, "service.namespace").
+			withTagPrefix(attrs, "telemetry.sdk"),
 	})
+	serviceInstanceIdentifier := fmt.Sprintf("urn:opentelemetry:namespace/%s:service/%s:serviceInstance/%s", serviceNamespace.AsString(), serviceName.AsString(), serviceInstanceId.AsString())
 	c.serviceInstances = append(c.serviceInstances, &Component{
-		fmt.Sprintf("urn:opentelemetry:namespace/%s:service/%s:serviceInstance/%s", serviceNamespace.AsString(), serviceName.AsString(), serviceInstanceId.AsString()),
+		serviceInstanceIdentifier,
 		ComponentType{
-			"service_instance",
+			"service-instance",
 		},
-		newData().
+		newComponentData().
 			withLayer("urn:stackpack:common:layer:containers").
 			withEnvironment(attrs).
 			withName(attrs, "service.instance.id").
 			withVersion(attrs, "service.version").
 			withTag(attrs, "service.namespace").
 			withTags(attrs),
 	})
+	c.addRelation(serviceIdentifier, serviceInstanceIdentifier, "provided by")
+	return true
+}
+
+func (c *ComponentsCollection) AddConnection(attrs *pcommon.Map) bool {
+	reqAttrs := make(map[string]string, 4)
+	for _, key := range []string{
+		"client",
+		"client_service.namespace",
+		"server",
+		"server_service.namespace",
+		"connection_type",
+	} {
+		value, ok := attrs.Get(key)
+		if !ok {
+			return false
+		}
+		reqAttrs[key] = value.AsString()
+	}
+
+	instanceId, ok := attrs.Get("client_service.instance.id")
+	var clientInstanceId string
+	if !ok {
+		clientInstanceId = reqAttrs["client"]
+	} else {
+		clientInstanceId = instanceId.AsString()
+	}
+	sourceId := fmt.Sprintf("urn:opentelemetry:namespace/%s:service/%s:serviceInstance/%s", reqAttrs["client_service.namespace"], reqAttrs["client"], clientInstanceId)
+
+	instanceId, ok = attrs.Get("server_service.instance.id")
+	var serverInstanceId string
+	if !ok {
+		serverInstanceId = reqAttrs["server"]
+	} else {
+		serverInstanceId = instanceId.AsString()
+	}
+	targetId := fmt.Sprintf("urn:opentelemetry:namespace/%s:service/%s:serviceInstance/%s", reqAttrs["server_service.namespace"], reqAttrs["server"], serverInstanceId)
+	c.addRelation(sourceId, targetId, reqAttrs["connection_type"])
```

Comment:
I'm pretty sure the `connection_type` is an empty string for anything that's a "normal" connection and not a database, messaging_system or virtual_node.

Reply:
ah, indeed.